DYNAMIC_GROUPS and NEW_GROUP_REQUEST parameters (#1060)
moq-wg/moq-transport: draft-ietf-moq-transport.md

The MOQT Object model is designed with the concept that the beginning of a Group
is a joint point, so in order for a subscriber to join a Track, it needs to
request an existing or future Group.  Different applications will have different
approaches for when to begin a new Group.

To join a Track at a past Group, the subscriber sends a SUBSCRIBE with Filter
Type `Largest Object` followed by a Joining FETCH (see {{joining-fetches}}) for
the intended start Group, which can be relative.  To join a Track at the next
Group, the subscriber sends a SUBSCRIBE with Filter Type `Next Group Start`.

#### Dynamically Starting New Groups

While some publishers will deterministically create new Groups, other
applications might want to only begin a new Group when needed.  A subscriber
joining a Track might detect that it is more efficient to request the Original
Publisher create a new group than issue a Joining FETCH.  Publishers indicate a
Track supports dynamic group creation using the DYNAMIC_GROUPS parameter
({{dynamic-groups}}).

One possible subscriber pattern is to SUBSCRIBE to a Track using Filter Type
`Largest Object` and observe the `Largest Location` in the response.  If the

> Object ID is below the application's threshold, the subscriber sends a FETCH for

Reviewer: ```suggestion
Object ID is below the application's threshold, the subscriber sends a joining FETCH for
```

Author: In the SUBSCRIBE/SUBSCRIBE_OK/FETCH flow -- it doesn't need to be a Joining FETCH, since you will know the correct Largest Object.  It could be a Joining FETCH if we want to document a PUBLISH/PUBLISH_OK/FETCH flow.

the beginning of the Group.  If the Object ID is above the threshold and the

> Track supports dynamic groups, it sends a SUBSCRIBE_UPDATE message with the

Reviewer: ```suggestion
Track supports dynamic groups, the subscriber sends a SUBSCRIBE_UPDATE message with the
```

NEW_GROUP_REQUEST parameter equal to the Largest Location's Group, plus one (see
{{new-group-request}}).

> Another possible subscriber pattern is to send a SUBSCRIBE with Filter Type

Reviewer: Can you add a bit more text on what this does?

> Another possible subscriber pattern is to send a SUBSCRIBE with Filter Type
> `Next Group Start` and NEW_GROUP_REQUEST equal to 0.

Author: ```suggestion
Another possible subscriber pattern is to send a SUBSCRIBE with Filter Type
`Next Group Start` and NEW_GROUP_REQUEST equal to 0.  The value of
DYNAMIC_GROUPS in SUBSCRIBE_OK will indicate if the publisher supports dynamic
groups. A publisher that does will begin the next group as soon as practical.
```

An endpoint that receives a filter type other than the above MUST close the
session with a `PROTOCOL_VIOLATION`.

### Joining an Ongoing Track

The MOQT Object model is designed with the concept that the beginning of a Group

> is a joint point, so in order for a subscriber to join a Track, it needs to

Reviewer 2: `a join point`?

> request an existing or future Group.  Different applications will have different

Reviewer 2: request an existing group or wait for a new one?

a relay that handles a downstream request that includes those Objects
re-requests them.

#### DYNAMIC GROUPS Parameter {#dynamic-groups}

The DYNAMIC_GROUPS parameter (parameter type 0x20) MAY appear in PUBLISH or
SUBSCRIBE_OK.  Values larger than 1 are a Protocol Violation.  When the value is
1, it indicates that the subscriber can send a NEW_GROUP_REQUEST parameter in

> PUBLISH_OK or SUBSCRIBE_UPDATE for this Track.

Author: @suhasHere I think you are saying a relay MUST store/forward this parameter to downstream subscribers, and implement the NGR logic?  This seems fine.

> PUBLISH_OK or SUBSCRIBE_UPDATE for this Track.

Reviewer 3: The original publisher is free to apply its logic on when and how to generate the new group. An application, say video conferencing app that supports non-static GOPs, will apply some hueristcs/batching logically at its end to decide when to generate a new group on the data stream.

> PUBLISH_OK or SUBSCRIBE_UPDATE for this Track.

Reviewer 4: +1 but a little vague and open ended

> 

Author: ```suggestion

Relays MUST preserve the value of this parameter received from an upstream
publisher in SUBSCRIBE_OK or PUBLISH when sending these messages to downstream
subscribers.

```

#### NEW GROUP_REQUEST Parameter {#new-group-request}

The NEW_GROUP_REQUEST parameter (parameter type 0x22) MAY appear in PUBLISH_OK,
SUBSCRIBE or SUBSCRIBE_UPDATE.  It is an integer representing the largest Group
ID in the Track known by the subscriber, plus 1. A value of 0 indicates that the
subscriber has no Group information for the Track.  A subscriber MUST NOT send
this parameter in PUBLISH_OK or SUBSCRIBE_UPDATE if the publisher did not
include DYNAMIC_GROUPS=1 when establishing the subscription.  A subscriber MAY
include this parameter in SUBSCRIBE without foreknowledge of support.  If the
publisher does not support dynamic Groups, it ignores the parameter in that
case.

> A Original Publisher that supports dynamic Groups that receives a
> NEW_GROUP_REQUEST with a value of 0 or a value larger than the current Group
> SHOULD end the current Group and begin a new Group as soon as practical.  The

Reviewer: ```suggestion
When an Original Publisher that supports dynamic Groups receives a
NEW_GROUP_REQUEST with a value of 0 or a value larger than the current Group,
it SHOULD end the current Group and begin a new Group as soon as practical.  The
```

Original Publisher MAY delay or ignore the NEW_GROUP_REQUEST subject to
implementation specific concerns, for example, acheiving a minimum duration for
each Group. The Original Publisher chooses the next Group ID; there are no

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Reviewer: I think it's ok to say that if a subscriber says their largest Group is X, and X has never been published, that's an error of some sort.

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Author: If we removed the group ID from the request, the downside is that the OP may have already begun the group needed by the subscriber and it will get a new request it is unable to deduplicate.  This will cause unnecessary increase in bandwidth to generate a new i-frame (even if the OP delays some minimum).  With the ID, either the relay or the OP can optimize away NGR that have already been satisfied.

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Reviewer 3: > I want to highlight this difference with #1031 - if a publisher is on group 99 and receives NGR=200, it does not need to make the next Group ID=200. It didn't seem right to give the subscribers that much control over track construction? If the subscriber is subscribed to this publisher, it can't have a largest group greater than 99, so it sent 200 because it's buggy or malicious. Otherwise it's some kind of multi-publisher scenario, in which case I'm not sure DYNAMIC_GROUPS make much sense.

I thunk it is not clear in #1031. But you are right in the intention with the either of the approaches. Subscriber doesn;t control the group Id generation, but just a new sync point that publisher can do .

case.

When an Original Publisher that supports dynamic Groups receives a
NEW_GROUP_REQUEST with a value of 0 or a value larger than the current Group,
it SHOULD end the current Group and begin a new Group as soon as practical.  The
Original Publisher MAY delay the NEW_GROUP_REQUEST subject to
implementation specific concerns, for example, acheiving a minimum duration for
each Group. The Original Publisher chooses the next Group ID; there are no
requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Relay Handling:

A relay that receives a NEW_GROUP_REQUEST for a Track without an active
subscription MUST include the NEW_GROUP_REQUEST when subscribing upstream.

A relay that receives a NEW_GROUP_REQUEST for an established subscription with a
value of 0 or a value larger than the Largest Group MUST send a SUBSCRIBE_UPDATE
including the NEW_GROUP_REQUEST to the publisher unless:

1. The Track does not support dynamic Groups
2. There is already an outstanding NEW_GROUP_REQUEST from this Relay with a
   greater or equal value

If a relay receives a NEW_GROUP_REQUEST with a non-zero value less than or equal

> to the Largest Group, it does not send a NEW_GROUP_REQUEST upstream.

Reviewer 5: There sort of is the ACK in the sense that all of the control messages are over a reliable transport but if some relay looses it, I agree but I don't see this as a problem. I think systems that use new groups at all are likely to pick a periodic update where they do a new group even if none were requested and it will be easy to see this problem in metrics.

> to the Largest Group, it does not send a NEW_GROUP_REQUEST upstream.

Author: See above comment regarding REQUEST_OK / REQUEST_ERROR as a solution to #1003

a relay that handles a downstream request that includes those Objects
re-requests them.

#### DYNAMIC GROUPS Parameter {#dynamic-groups}

> The DYNAMIC_GROUPS parameter (parameter type 0x20) MAY appear in PUBLISH or

Reviewer 4: By that, subsequent RFC/drafts to this protocol will likely add more params. Those params will end up going to the publisher, no different than new group request. Why go to this level just for new group? Sounds like the live-stream folks are interested in param/update filtering.  Basically, notify me of X but not Y.  Do not include params that I do not support.  This is classic capabilities exchange on params.

SUBSCRIBE_OK.  Values larger than 1 are a Protocol Violation.  When the value is
1, it indicates that the subscriber can send a NEW_GROUP_REQUEST parameter in
PUBLISH_OK or SUBSCRIBE_UPDATE for this Track.

> Relays MUST preserve the value of this parameter received from an upstream

Reviewer 3: Since the groupId is latching point and avoids using timers configured across all the relays. In the current proposal, the timers or such logic is moved only to the OP.

> Relays MUST preserve the value of this parameter received from an upstream

Reviewer 4: Recall the implementation will apply this at the publisher edge relay only. Timers for such can be negotiated via parameters for such.

> Relays MUST preserve the value of this parameter received from an upstream

Author: > Why?

This param indicates to both relays and subscribers that NGR is allowed.  It needs to propagate anywhere the Track goes.

> Relays MUST preserve the value of this parameter received from an upstream

Reviewer 4: @fluffy , I believe that makes an assumption that the inter-relay protocol is MOQT hop-by-hop. My understanding is that MOQT is not defining inter-relay protocol.  Has that changed? 

My two-cents are that this is a publisher capability exchange for subscribers, which is the first that needs to be conveyed to every subscriber via `SUBSCRIBE_OK`.  This is new and is pretty significant. If MOQT is going to start advertising capabilities of publishers to subscribers, then wouldn't there be more in the very near future? Either by this draft or by subsequent follow-up drafts that add publisher capabilities (aka params) that need to be made available to subscribers?  I'm supported of having that as a feature, but I do feel that this isn't solving how we sustain one or more capabilities of a publisher to subscribers. How we do that also needs to take into account multi-publisher for the same track.  We can certainly do a more efficient method to exchange and provide publisher capabilities (multi-publisher supported) to subscribers that join anywhere in the infrastructure, regardless of how many relays are involved.  IMO, the param needs to be marked as **transitive** vs **non-transitive**.  If the param is transitive, then the parm added to `SUBSCRIBE_OK` of subscribe. Subscriber will get the param. If non-transitive, it stays local and is not sent in `SUBSCRIBE_OK` to client subscribers.

> Relays MUST preserve the value of this parameter received from an upstream

Author: > How we do that also needs to take into account multi-publisher for the same track.

I think in this PR, if there are two publishers with different DYNAMIC_GROUPS values for the same Track, I think it still works: one of them would get the NGR and one wouldn't.  I'm not sure that's likely though in the cases we care about.

publisher in SUBSCRIBE_OK or PUBLISH when sending these messages to downstream
subscribers.

#### NEW GROUP_REQUEST Parameter {#new-group-request}

The NEW_GROUP_REQUEST parameter (parameter type 0x22) MAY appear in PUBLISH_OK,
SUBSCRIBE or SUBSCRIBE_UPDATE.  It is an integer representing the largest Group
ID in the Track known by the subscriber, plus 1. A value of 0 indicates that the
subscriber has no Group information for the Track.  A subscriber MUST NOT send
this parameter in PUBLISH_OK or SUBSCRIBE_UPDATE if the publisher did not

> include DYNAMIC_GROUPS=1 when establishing the subscription.  A subscriber MAY

Author: If the state you are referring to is the opt-in parameter, I think the wg felt strongly this was needed.

> include DYNAMIC_GROUPS=1 when establishing the subscription.  A subscriber MAY

Reviewer 4: The main issue here is that the publisher may not support NGR and the subscriber is unaware of that, resulting in repeated attempts for NGR from the subscriber to no avail. This causes a lot of churn for both the relays and publisher. This capability being advertised sounds like a good idea to notify the subscriber, but IMHO, this opens a bit of pandora's box. NGR is just one capability, there will likely be many more capabilities that are similar to NGR over time.  MOQT should just facilitate the exchange of capabilities. If the capabilities were made available to the subscriber, then the subscriber can act accordingly. If the edge relay can provide in-depth filtering, then it can filter a subscriber sending NGR or other params that shouldn't result in notifying the publisher. I see this as edge relay function, which requires publisher capabilities (aka transitive params for one or more publishers to the same track) to be known.

> include DYNAMIC_GROUPS=1 when establishing the subscription.  A subscriber MAY

Author: > The key issue here is not if the original publisher supports it but if all the relays on the path support it.

The relay handling described here is a MUST for Relays, so all Relays on the path do support it.  It's really a question if the Original Publisher supports it.

> An issue here is that the publisher may not support NGR and the subscriber is unaware of that

That's why DYNAMIC_GROUPS needs to be conveyed in SUBSCRIBE_OK/PUBLISH.  If we feel like Params is not a good fit here, consider it a track-level Extension (see #1176).

Regarding Generic Capabilities Exchange:

I think additional Publisher -> Subscriber capabilities can be conveyed in future versions without any additional help from Relays.  However, Subscriber -> Original Publisher messages need some kind of relay support which is tricky.  I'd like to see a concrete proposal and PR to fully understand.

include this parameter in SUBSCRIBE without foreknowledge of support.  If the
publisher does not support dynamic Groups, it ignores the parameter in that
case.

When an Original Publisher that supports dynamic Groups receives a
NEW_GROUP_REQUEST with a value of 0 or a value larger than the current Group,
it SHOULD end the current Group and begin a new Group as soon as practical.  The

> Original Publisher MAY delay or ignore the NEW_GROUP_REQUEST subject to

Author: ```suggestion
Original Publisher MAY delay the NEW_GROUP_REQUEST subject to
```

implementation specific concerns, for example, acheiving a minimum duration for
each Group. The Original Publisher chooses the next Group ID; there are no

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Reviewer 4: Dampening IMO is required. Bool is totally fine and group id only complicates things with reconciliation between other subscriber requests. The net result is to start a new group, despite whatever the group id is. It’s a classic synchronization request to start anew.

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Reviewer 5: Say we have 100k people joining a company meeting in the first 2 minutes of the call. First person joins, great and publisher sends I frame as group=1. Now the next 100k-1 people are going to request a new group. Let say in the first 2 seconds , 10 k people join, and in the next 2 seconds another 10k people join and that the original publisher is configure not to do an I frame more often than every 2 seconds. 

The goal is for the relay to end up agregating most of the requests so the original publisher only gets a new group request shortly after each time it starts a new group ( every 2 seconds in this case )

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Author: Ok, I see the problem:  If current group is 10 and someone sends NGR=1000000, the OP ignores it and keeps sending the current group.  Then legit subscribers cannot send NGR=11.

My vision is that #1182 / #1159 will soon positively or negatively acknowledge SUBSCRIBE_UPDATE.  In that case, the OP can send a REQUEST_ERROR for invalid NGR, which could clear the state at the relay.

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Author: > A new subscriber can still send 11 and it will work .

How?  11 < 10000000 and the relay has not seen a new group to clear the old request.

> requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

Reviewer 3: The OP supporting NGR cannot ignore the request. It will send the next group based on its application logic (may be delayed) and that should clear the one for 10000000.  

Yes something like this can cause subscribers to not get new group for a while, but it cannot block things indefinitely unless the OP decides to never start a new group in which case the expected behavior on the subscriber is same with NGR carrying larger or smaller number.  This seems like application bug and not to be dealt at moq level and I think i did clarify this to Tim as well.

Relay Handling:

> A relay can have at most one outstanding NEW_GROUP_REQUEST per Track at a time.

Author: The key difference is that if a Relay forwards NGR 10, it won't forward NGR 11 until after it sees group 10 (or higher).  This prevents misbehaving clients from sending NGR 1...MAXINT, flooding the OP.

Can you explain the race conditions you see in detail?

> A relay can have at most one outstanding NEW_GROUP_REQUEST per Track at a time.

Author: @fluffy in case you missed my question here

After sending a NEW_GROUP_REQUEST upstream, the request is considered

> outstanding until the Largest Group increases.

Reviewer 4: Makes sense to dampen the flood of new group requests, but this method is broken when the group never changes. Based on this text, new group will never be sent again if the group doesn't increment... The APP may decide to ignore the request based on timing/whatever.  It may then never increment the group id. Had a new group request been sent again later it might have been accepted and resulted in a new group. 

```suggestion
complete. Subsequent NEW_GROUP_REQUEST received SHOULD be aggregated to a single request to the publisher that will be delayed by CLIENT_SETUP parameter TBD milliseconds from when the last request was sent. 
```

Author: I think you are right. The OP shouldn't _ignore_ an NGR, but it can arbitrarily delay it.

Reviewer 6: No, I am strongly opposed to any message that would force an Origin Publisher to create a New Group upon request, even delayed.

It was very clear since the beginning (and we collectively agreed) that this New Group Request is a great feature for video conferencing scenario but it is dangerous in any massive live streaming use case, so it must not be a forceful message. We agreed to debate on this topic at the condition that the Origin Publisher can decide to not apply a new group.

An Origin Publisher for the Superbowl or for moon landing does not want to pay any single attention to NGR messages.

Reviewer 6: Not really. The OP should be free to decide whether it wants to create a new group, to delay this creation, or to ignore. There are cases where it makes sense to create a group (dynamic frames, small population, many resources...) and ten minutes later, it does not make sense any longer (static frames, large population, less resources...).
Imho, subscribers are free to send a NGR, and, well, they may get a new group soon, or not, ... but they have to understand that there is no "right to get a new group". The OP has a much wider understanding of the configuration and he may not fulfill all requests.

Author: The only way to truly "ignore" an NGR is to _never_ make a new group.  If we retain the algorithm described here, an OP that silently drops an NGR and never starts a new group might never get another NGR.  Perhaps we can leave "ignore" in as a valid action with that caveat?

We can also fix this is by relaxing the restriction at the relay (as Cullen suggests), at the expense of losing some of the protection offered by the "one NGR at a time" rule.

A relay that receives a NEW_GROUP_REQUEST for a Track without an active
subscription MUST include the NEW_GROUP_REQUEST when subscribing upstream.

A relay that receives a NEW_GROUP_REQUEST for an established subscription with a
value of 0 or a value larger than the Largest Group MUST send a SUBSCRIBE_UPDATE

> including the NEW_GROUP_REQUEST to the publisher unless:

Reviewer 4: The below list will likely only be known by the origin relay.  The downstream relays do not have direct access to the publisher and the negotiated params for that publisher connection.  This is especially the case with multi-publisher. 

The text should say origin relay does this.

Reviewer 5: Tim, it seems each relay would individually know this with no need to synchronize it across the relay mesh

Author: > "since we forward DYNAMIC_GROUPS" is an assumption for inter-relay communication.

In MOQT we describe a single Relay.  If you have a mesh of Relays that act as a single Relay, then it needs to conform to the requirements in the spec.  So yes, in a way.

> MOQT is not defining inter-relay protocol

100% agree, but see above.  You can use any inter-relay protocol you want.  The requirement is if you are a Relay, you include DYNAMIC_GROUPS in the SUBSCRIBE_OK/PUBLISH if you received it.

> We support lingering subscribers, where subscribers join before the publisher.

I'm pretty sure sending the SUBSCRIBE_OK before having an upstream subscription is now disallowed.  If the subscriber shows up first it should use SUBSCRIBE_NAMESPACE, and then it will get a PUBLISH.

> Per the text in the PR, this would result in flooding updates to the publisher(s) regardless of less than or equal to last NGR.

I'm not sure restricting 0 to SUBSCRIBE will prevent a DoS of the OP, since an attacker can just send SUBSCRIBE.  But also, if SUBSCRIBE has 0 but there's already a subscription, the relay needs to send SUBSCRIBE_UPDATE with NGR=0, so I don't think we can restrict it.

1. The Track does not support dynamic Groups

> 2. There is already an outstanding NEW_GROUP_REQUEST

Author: ```suggestion
2. There is already an outstanding NEW_GROUP_REQUEST from this Relay
```

If a relay receives a NEW_GROUP_REQUEST with a non-zero value less than or equal

> to the Largest Group, it does not send a NEW_GROUP_REQUEST upstream.

Reviewer 5: If the NGR 15 had completed at the point the NGR 12 happened, there would not be an outstanding and at that point I thought it would send the NGR 12. I thought that was what was being proposed here but perhaps I had it wrong. I guess I might be unclear on what outstanding transaction means

> to the Largest Group, it does not send a NEW_GROUP_REQUEST upstream.

Author: @fluffy 

> If the NGR 15 had completed at the point the NGR 12 happened, there would not be an outstanding and at that point I thought it would send the NGR 12.

This is how it works. Per my example above:

```
Relay Largest = 10
S1->R: NGR 15
R->P: NGR
P->R: Object, Group=11
S2->R: NGR 12
R->P: NGR
```

Conceptually the NGR has _no group value_ when received by the OP (relays do include it in case next upstream is a relay).  See this bit:

>  The Original Publisher chooses the next Group ID; there are no requirements that it be equal to the NEW_GROUP_REQUEST parameter value.

> I guess I might be unclear on what outstanding transaction means

The definition is on line 1641:

> After sending a NEW_GROUP_REQUEST upstream, the request is considered outstanding _until the Largest Group increases_.

Can you articulate why you think this is a bad design?  And why this is a necessary requirement?

> it should also not send it if it has already sent a new group request that is equal too or larger than this

> to the Largest Group, it does not send a NEW_GROUP_REQUEST upstream.

Author: Spoke offline with @fluffy.  The scenario where the "one outstanding" request does not work is as follows:

Largest Group seen in parens
```
C1 (22) -> R1 (22) -> P (24)
C2 (24) -> R2 (24) -/
```

R1 and C1 got behind.  Now C2 fails over from R1 to C1.

```
C1 (22) -> R1 (22) -> P (24)
C2 (24) -/
```

C1 sends NGR 23
C2 sends NGR 25

Under the original algorithm, the OP would discard NGR 23, because it has published 24, but R1 would discard NGR 25, because NGR 23 is outstanding (until the an object from 24 reaches R1).

The tradeoff is that a buggy or malicious client can send an unlimited number of NEW_GROUP_REQUESTs to the OP.  Cullen pointed out that a client can already send an unlimited number of FETCH requests that reach the OP as well.

If the EXPIRES parameter is 0 or is not present in a message, the subscription
does not expire or expires at an unknown time.

#### DYNAMIC GROUPS Parameter {#dynamic-groups}

> The DYNAMIC_GROUPS parameter (parameter type 0x20) MAY appear in PUBLISH or
> SUBSCRIBE_OK.  Values larger than 1 are a Protocol Violation.  When the value is
> 1, it indicates that the subscriber can send a NEW_GROUP_REQUEST parameter in
> PUBLISH_OK or SUBSCRIBE_UPDATE for this Track.

Author: ```suggestion
The DYNAMIC_GROUPS parameter (parameter type 0x20) MAY appear in PUBLISH or
SUBSCRIBE_OK.  Values larger than 1 are a Protocol Violation.  When the value is
1, it indicates that the subscriber can request the Original Publisher to start
a new Group by including the NEW_GROUP_REQUEST parameter in PUBLISH_OK or
SUBSCRIBE_UPDATE for this Track.
```

Relays MUST preserve the value of this parameter received from an upstream
publisher in SUBSCRIBE_OK or PUBLISH when sending these messages to downstream
subscribers.

#### NEW GROUP_REQUEST Parameter {#new-group-request}

The NEW_GROUP_REQUEST parameter (parameter type 0x22) MAY appear in PUBLISH_OK,
SUBSCRIBE or SUBSCRIBE_UPDATE.  It is an integer representing the largest Group
ID in the Track known by the subscriber, plus 1. A value of 0 indicates that the
subscriber has no Group information for the Track.  A subscriber MUST NOT send
this parameter in PUBLISH_OK or SUBSCRIBE_UPDATE if the publisher did not
include DYNAMIC_GROUPS=1 when establishing the subscription.  A subscriber MAY

> include this parameter in SUBSCRIBE without foreknowledge of support.  If the

Reviewer: I don't think this behavior is a good idea, since it's a silent failure.

Author: It's not a silent failure -- SUBSCRIBE_OK will lack DYNAMIC_GROUPS, and the subscriber will know that they need to FETCH or wait.

If the EXPIRES parameter is 0 or is not present in a message, the subscription
does not expire or expires at an unknown time.

#### DYNAMIC GROUPS Parameter {#dynamic-groups}

The DYNAMIC_GROUPS parameter (parameter type 0x20) MAY appear in PUBLISH or
SUBSCRIBE_OK.  Values larger than 1 are a Protocol Violation.  When the value is
1, it indicates that the subscriber can request the Original Publisher to start
a new Group by including the NEW_GROUP_REQUEST parameter in PUBLISH_OK or
SUBSCRIBE_UPDATE for this Track.

Relays MUST preserve the value of this parameter received from an upstream
publisher in SUBSCRIBE_OK or PUBLISH when sending these messages to downstream
subscribers.

#### NEW GROUP_REQUEST Parameter {#new-group-request}

The NEW_GROUP_REQUEST parameter (parameter type 0x22) MAY appear in PUBLISH_OK,
SUBSCRIBE or SUBSCRIBE_UPDATE.  It is an integer representing the largest Group
ID in the Track known by the subscriber, plus 1. A value of 0 indicates that the

> subscriber has no Group information for the Track.  A subscriber MUST NOT send

Reviewer: Is it valid to combine this with Forward=0?

Author: It seems a little mean to send NGR Forward=0, but not sure if we want to make it an error.

Reviewer: I'm ok with leaving this as is, but it is odd.

> this parameter in PUBLISH_OK or SUBSCRIBE_UPDATE if the publisher did not

Author: Good catch, can you suggest the change you want, if any?  Just allow it in SUBSCRIBE_UPDATE and the publisher can ignore?

> include DYNAMIC_GROUPS=1 when establishing the subscription.  A subscriber MAY

Reviewer: ```suggestion
include DYNAMIC_GROUPS=1 when establishing the subscription.  As such, if a
SUBSCRIBE_UPDATE is sent prior to receiving a SUBSCRIBE_OK, the SUBSCRIBE_UPDATE
MUST NOT include a NEW_GROUP_REQUEST param. A subscriber MAY
```

include this parameter in SUBSCRIBE without foreknowledge of support.  If the

> publisher does not support dynamic Groups, it ignores the parameter in that

Reviewer: Is this original publisher?  Because this is required for relays, correct?

```suggestion
original publisher does not support dynamic Groups, it ignores the parameter in that
```

Reviewer: I guess my point is this is a required feature in the core draft, so it's really the original publisher that's going to support this, correct?  And anyone downstream of them is just reflecting that value?
